experimental/transliterator_parser/src/compile.rs

```diff
@@ -2,8 +2,1243 @@
+//! accounted for.
+//!
+//! ## Pass 2
+//! Encoding of the zero-copy data struct.
+//!
+//! To encode conversion rules into `str`s, we use the previously described encoded `VarTable`
+//! indices. Because we know the lengths of each special construct list (in the form a `VZV`)
+//! from the first pass, we can store the offsets for each special construct list (i.e., the sum of
+//! the lengths of the previous lists) while encoding the conversion rules, and incrementing the
+//! offset of a given special construct when we encode an element. The precomputed lengths mean we
+//! never overflow into the indices of the following `VZV`.
+
+// more (data struct compatible) runtime optimization opportunities:
+// - deduplicate special constructs ($a = hello; $b = hello; should only generate one hello element)
+//   - especially important for equivalent unicodesets
+// - inline single-use variables ($a = x; $a > b; => x > b;)
+// - replace uses of single-element variables with the element itself ($a = [a-z]; $a > a; => [a-z] > a;)
+// - flatten single-element sets into literals ([a] > b; => a > b;)
+
+/*
+Encoding example:
+
+    $b = bc+ ;
+    $a = [a-z] $b ;
+    $a > ;
+
+b-data.counts: 1 compound (the definition itself), 1 quantifier plus (c+)
+b-data.used_vars: -
+
+a-data.counts: 1 compound (the definition itself), 1 unicodeset ([a-z])
+a-data.used_vars: b
+
+forward-data.counts: 0 (rules are inlined)
+forward-data.used_vars: a
+
+when collecting the counts (for forward) at the end, we sum over all counts of the transitive
+dependencies of forward (using used_vars), and add the counts of forward itself.
+we also compute the transitive closure of used variables.
+this gives us the `Pass1Result`:
+forward-data.counts: 2 compound, 1 quantifier plus, 1 unicodeset
+forward-data.used_vars: a, b
+
+this `Pass1Result` we give to Pass2, which will produce something like this:
+(note that the integer-indexed maps shown here are only semantic, in actuality the indices are implicit,
+as described in the zero-copy format, and the maps here are just arrays)
+
+    VarTable {
+        compounds: {
+            0: "b<2>", // b's definition, bc+
+            1: "<3><0>", // a's definition, [a-z] $b
+        },
+        quantifier_kleene_plus: {
+            2: "c", // c+
+        },
+        unicode_sets: {
+            3: <the set of a..z>, // [a-z]
+        }
+    }
+    Rules: [
+        {
+            source: "<1>", // $a
+            target: "",
+        }
+    ]
+*/
+
+use crate::parse;
+use crate::parse::{ElementLocation as EL, HalfRule, QuantifierKind};
+use parse::Result;
+use parse::PEK;
+use std::collections::{HashMap, HashSet};
+
+enum SingleDirection {
+    Forward,
+    Reverse,
+}
+
+/// The number of elements for each `VZV` in the `VarTable`.
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq)]
+struct SpecialConstructCounts {
+    num_compounds: usize,
```

Comment:
I think that's a leftover from another design where I had planned to make other special constructs refer to `vartable.compounds` instead of inlining the `str`. I agree with the change and filed an issue: #3850